datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java

```diff
@@ -170,17 +173,35 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
 
     DatastreamGroup dg = partitionsMetadata.getDatastreamGroup();
 
+    // Pre-process the targetAssignment to filter out partitions that are supposed to move to the target instance that
+    // they are already on to avoid unnecessary partition movements. That is, if partition1 is currently on instance1
+    // and is supposed to move to instance1, remove it from the targetAssignment as this should be a no-op.
+    Map<String, Set<String>> preprocessedTargetAssignment = new HashMap<>();
+    targetAssignment.forEach((instance, partitions) -> {
+      Set<DatastreamTask> tasks = currentAssignment.get(instance);
+      Set<String> partitionsAcrossAllDatastreamTasks = tasks.stream().filter(dg::belongsTo)
+          .map(DatastreamTask::getPartitionsV2).flatMap(Collection::stream).collect(Collectors.toSet());
+
+      Set<String> updatedTargetPartitionList = new HashSet<>(partitions);
+      updatedTargetPartitionList.removeAll(partitionsAcrossAllDatastreamTasks);
+
+      preprocessedTargetAssignment.computeIfAbsent(instance,
+          val -> updatedTargetPartitionList.isEmpty() ? null : updatedTargetPartitionList);
+    });
+
+    LOG.info("The preprocessed targetAssignment: {}", preprocessedTargetAssignment);
+
     Set<String> allToReassignPartitions = new HashSet<>();
-    targetAssignment.values().forEach(allToReassignPartitions::addAll);
+    preprocessedTargetAssignment.values().forEach(allToReassignPartitions::addAll);
```

Comment:
addAll with throw NullPointerException for the value set to null. The value is set to null for the instance where there is no movement.

Reply:
In the previous step, where we create `preprocessedTargetAssignment`, the `computeIfAbsent` does not add the key to the map if the computed value is null. I've modified this to filter out null values anyways just to be safe, but using `addAll` here without null checks should be safe.

Code that updates `preProcessedTargetAssignment`:
```
      preProcessedTargetAssignment.computeIfAbsent(instance,
          val -> updatedTargetPartitionList.isEmpty() ? null : updatedTargetPartitionList);
```

Comments on `computeIfAbsent`:
>      * If the specified key is not already associated with a value (or is mapped
>     * to {@code null}), attempts to compute its value using the given mapping
>     * function and enters it into this map unless {@code null}.

Reply:
Make sense. You can ignore this comment. 

Feel free to ignore this suggestion. Just wondering should we make this piece more simple.
```
if (!updatedTargetPartitionList.isEmpty()) {
preProcessedTargetAssignment.put(instance, updatedTargetPartitionList);
}

Reply:
I sort of prefer `computeIfAbsent` here, so will leave this as is.